### PR TITLE
Fix index problem with 2-body Jastrow optimization

### PR DESCRIPTION
**Pull request by markdewing**
Addresses #2814 (at least partially)

Add a unit test that will fail in one case where the offsets are not computed correctly.

Fix the offset calculation in the case the first function does not have any optimizable parameters.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
laptop

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h [hunk 1]
```diff
@@ -139,7 +145,18 @@ class DiffTwoBodyJastrowOrbital : public DiffWaveFunctionComponent
...
+      int varoffset = -1;
+      for (int i = 0; i < myVars.size(); i++)
+      {
+        varoffset = myVars.Index[i];
+        if (varoffset != -1)
+        {
+          break;
+        }
+      }
+
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h:159]: If we have 3 distinct sets of coefficients (uu/ud/dd) and only the middle one is optimization. The code is still potentially wrong? Let me know your answer or your guess.

File: src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h [hunk 2]
```diff
@@ -139,7 +145,18 @@ class DiffTwoBodyJastrowOrbital : public DiffWaveFunctionComponent
...
       OffSet.resize(F.size());
-      int varoffset = myVars.Index[0];
+
+      // Find first active variable for the starting offset
+      int varoffset = -1;
+      for (int i = 0; i < myVars.size(); i++)
+      {
+        varoffset = myVars.Index[i];
+        if (varoffset != -1)
+        {
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h:155]: Please remove {} around break to save two lines.

File: src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h (hunk 1 above)
Comment (markdewing) [reply src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h:157]: I modified the unit test for this case, and it still passed.

File: src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h (hunk 2 above)
Comment (markdewing) [reply src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h]: fixed

File: src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h (hunk 1 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h:157]: The unit test is only limited to check indices. So it is guaranteed up-to what is being checked. Not necessary a real J2 optimization will work. right?

Comment (markdewing) [reply src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h:157]: Yes, that's correct - this only fixes setting OffSet variable.  Shiv's PR had another change that I need to look at.

Comment (ye-luo) [reply src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h:157]: I see thanks.

**Review by prckent — APPROVED** [review]
LGTM

I changed the title since it wouldn't surprise me if this code pattern is repeated in other wavefunction components.

**Pull request merged**